### DISCUSSION TITLE
Can't access local network devices

### DISCUSSION THREAD

**Discussion by Iandesuu** [pivpn/pivpn#1716] [Support]
Hello !
It's been a while since I have my raspberry pi 4 running as a home VPN. Hoewever, I never managed to access devices such as local Hard drives etc in my home's local network through the VPN.

I'm using the Wireguard.

Is there any way to access it ?

Thanks.

**Comment by orazioedoardo** [comment]
What's the subnet class of the PiVPN network, as well the subnet class of the network you are connecting from? Are they the same?

**Reply by Iandesuu** [reply]
Thanks for the reply!
Subnet class of my PiVPN was 24.
And on my router page it's written ***(IP ADRESS)*** / 23 for Lan and /24 for Wan.
But I'm using Lan so I think that they don't have the same subnet class.

So how can I change it ?

**Reply by orazioedoardo** [reply]
Actually what’s the network? Like 192.168.0.0/24

**Reply by Iandesuu** [reply]
Sure! it's written that it's 192.168.0.1/24 for my Lan network.

**Reply by orazioedoardo** [reply]
Same network for both LANs?

**Reply by Iandesuu** [reply]
I found this on my raspi(piVPN):

![image](https://user-images.githubusercontent.com/113456382/232247207-902eaeef-6abd-489b-8b04-9d5365611588.png)

**Reply by orazioedoardo** [reply]
That's but where are you connecting to the VPN from?

**Reply by Iandesuu** [reply]
home to home but accessing from another country (France -> Japan) 😅

**Reply by orazioedoardo** [reply]
Another network then. What I'm asking, is the France LAN the same as the Japan PiVPN LAN (192.168.0.0/24)?

**Reply by Iandesuu** [reply]
How can I check that ?

And it's 192.168.0.113/24 instead of 192.168.0.0/24 right?

**Reply by orazioedoardo** [reply]
The Raspberry has IP 192.168.0.113/24 which means the network is 192.168.0.0/24. As for checking the other network, look in the network settings of the client devices.

**Reply by Iandesuu** [reply]
![image](https://user-images.githubusercontent.com/113456382/232249141-496bcc27-adf8-413d-84da-5287e08940ac.png)


looking at this, it looks like that it's not 192.168.0.0/24

**Reply by orazioedoardo** [reply]
Not the wireguard network but the local network (wifi/ethernet)

**Reply by Iandesuu** [reply]
Here is it:

![image](https://user-images.githubusercontent.com/113456382/232249577-5b3b0972-a5a2-440b-a544-c52ea1a3573d.png)

**Reply by orazioedoardo** [reply]
Different (192.168.1.0/24) so it won't conflict. Can you post `pivpn -d`? Note that you don't need to redact the virtual network address, noone can do anything with it.

**Reply by orazioedoardo** [reply]
I’ve deleted your comment because endpoint wasn’t censored. Anyway, you can run a packet capture on the network using tcpdump. Install the `tcpdump` package, run `sudo tcpdump -n net 192.168.0.0/24` and try to connect to a device on the PiVPN LAN via VPN. You should see some live output on the terminal.

**Reply by Iandesuu** [reply]
Thank you very much. I run the command after that I installed and it showed me a bunch of thing so quickly so I stopped the thing but what should I watch at?

**Reply by orazioedoardo** [reply]
Ok so it’s likely picking up port 22 if you are connected via SSH. Try a single IP `sudo tcpdump -n host 192.168.0.1` and try to connect to 192.168.0.1 while on VPN

**Reply by Iandesuu** [reply]
When I tried to connect to 192.168.0.1, I saw that the live output was moving rapidly. And this brought me to the router's web-based setup page(JP), which I alredy had acces before.

**Reply by orazioedoardo** [reply]
Looks like LAN access works.

**Reply by Iandesuu** [reply]
Right. however, I still can't see my devices on my local network that is normally visible. the "LAPTOP-IAN" is the laptop which i'm using to connect through VPN.

![image](https://user-images.githubusercontent.com/113456382/232253662-1bff76e1-b982-477c-b629-94ec59cc3f0c.png)

**Reply by orazioedoardo** [reply]
>  I still can't see my devices on my local network that is normally visible.

Please clarify which LAN you are referring to.

**Reply by Iandesuu** [reply]
I'm trying to connect to "192.168.0.4"

![image](https://user-images.githubusercontent.com/113456382/232286851-3e258047-2c0e-4f93-bb84-88e1ffd33c41.png)

**Reply by orazioedoardo** [reply]
Devices likely do not appear in file browser due to the VPN not forwarding multicast DNS, however connecting to an IP address you know should work.

**Reply by Iandesuu** [reply]
Yes, I just tried to ping and got a reply. So is there any way to enable the forwarding multicast DNS ?

**Reply by Iandesuu** [reply]
I found this : 
Is this right? (Wroten by ChatGPT)

To enable multicast DNS (mDNS) forwarding over the VPN, you will need to configure your VPN server to support it. This process will vary depending on the VPN software you are using.

One common way to enable mDNS forwarding over VPN is to use the Avahi daemon, which is a free zero-configuration networking (zeroconf) implementation that allows devices to automatically discover and connect to each other on a local network. Avahi includes support for mDNS and can be used to advertise network services over the VPN.

To enable mDNS forwarding over VPN using Avahi, you can follow these steps:

Install the Avahi daemon on your VPN server. On most Linux distributions, you can do this by running the command sudo apt-get install avahi-daemon.

Configure the Avahi daemon to advertise the network services you want to make available over the VPN. This will typically involve creating service definition files in the /etc/avahi/services directory. For example, to advertise an SSH service, you can create a file named ssh.service with the following contents:

```
<?xml version="1.0" standalone='no'?><!--*-nxml-*-->
<!DOCTYPE service-group SYSTEM "avahi-service.dtd">

<service-group>
  <name replace-wildcards="yes">%h SSH</name>
  <service>
    <type>_ssh._tcp</type>
    <port>22</port>
  </service>
</service-group>
```

This file defines an mDNS service for SSH on port 22.

Edit the Avahi daemon configuration file (/etc/avahi/avahi-daemon.conf) to enable mDNS forwarding over VPN. You can do this by uncommenting and setting the enable-dbus and disallow-other-stacks options as follows:

```
enable-dbus=yes
disallow-other-stacks=yes
```
These options enable the Avahi daemon to listen for mDNS queries on the VPN interface and to prevent other network stacks from interfering with mDNS.

Restart the Avahi daemon to apply the changes: sudo service avahi-daemon restart
After completing these steps, devices connected to the VPN should be able to discover and connect to mDNS services advertised by the VPN server, such as SSH or HTTP services.

**Reply by orazioedoardo** [reply]
> Is this right? (Wroten by ChatGPT)

Lol that's a thing now. I haven't checked nor tried those steps but it looks like the configuration will **advertise the SSH service over the VPN**, however you want to **advertise the LAN** over VPN and that doesn't work over WireGuard because it's a "routed VPN". You would need to use OpenVPN in TAP mode "bridged VPN" but that's not supported by PiVPN.

https://openvpn.net/community-resources/determining-whether-to-use-a-routed-or-bridged-vpn/

**Reply by Iandesuu** [reply]
oh... wow. So It has to be completely another thing then. I don't think that I would be able to set up OpenVPN in tap mode manually.
But in case I really need it, can you tell me if it is possible to install and run open VPN tap mode with PiVPN at the same time? 
because what I saw was that we can't use it on Android etc. 
So like normal uses -> Wireguard
file share -> OpenVPN (tap mode)

**Reply by orazioedoardo** [reply]
OpenVPN in TAP mode (if you can setup yourself) and PiVPN WireGuard should not conflict.

**Reply by Iandesuu** [reply]
And a final question, Is tap mode can be used as VPN just like as wireguard but with file sharing?
And if I change to tap mode, I will be able to acces to the VPN(just internet, not file sharing) from my android Phone ?

**Reply by orazioedoardo** [reply]
> And a final question, Is tap mode can be used as VPN just like as wireguard but with file sharing?

File sharing itself does not need TAP mode, as I said connecting to a service on the LAN works fine if you know the address of the server or computer. What doesn't work is automatic discovery. 

> And if I change to tap mode, I will be able to acces to the VPN(just internet, not file sharing) from my android Phone ?

Android does not support TAP mode.

**Reply by Iandesuu** [reply]
> File sharing itself does not need TAP mode

I tried everything like putting the exact adress of the file on the file explorer but that doesn't work.


> Android does not support TAP mode.

So I will never change to tap mode lol
